SEIR/NPI/Stacked.py

```diff
@@ -1,58 +1,67 @@
-        reduction_lists = {}
-        for param in REDUCE_PARAMS:
-            reduction_lists[param] = []
-
-        self.sub_npis = []
-
-        self.intervention_name = npi_config.name
-
-        for scenario in npi_config["scenarios"]:
+        # the confuse library's config resolution mechanism makes slicing the configuration object expensive; instead,
+        # just preload all settings
+        settings_map = global_config["interventions"]["settings"].get()
+        for scenario in npi_config["scenarios"].get():
             # if it's a string, look up the scenario name's config
-            if isinstance(scenario.get(), str):
-                scenario_npi_config = global_config["interventions"]["settings"][scenario.get()]
-                if not scenario_npi_config.exists():
+            if isinstance(scenario, str):
+                settings = settings_map.get(scenario)
+                if settings is None:
                     raise RuntimeError(f"couldn't find scenario in config file [got: {scenario}]")
             else:
                 # otherwise use the specified map as the config
-                scenario_npi_config = scenario
-
-            sub_npi = NPIBase.execute(npi_config=scenario_npi_config, global_config=global_config, geoids=geoids, loaded_df = loaded_df)
-            self.sub_npis.append(sub_npi)
-
-
-        self.reductions = {}
-        for param in REDUCE_PARAMS:
-            self.reductions[param] = 1 - functools.reduce(lambda a,b : a * (1 - b.getReduction(param)), self.sub_npis, 1)
+                settings = scenario
+
+            # via profiling: faster to recreate the confuse view than to fetch+resolve due to confuse isinstance checks
+            scenario_npi_config = confuse.RootView([settings])
+            sub_npi = NPIBase.execute(npi_config=scenario_npi_config, global_config=global_config, geoids=geoids,
+                                      loaded_df=loaded_df)
+            for param in REDUCE_PARAMS:
+                reduction = sub_npi.getReduction(param)
+                if reduction is not None:
+                    self.reductions[param] = (self.reductions[param] * (1 - reduction)
+                                              if self.reductions[param] is not None else reduction)
+
+                # FIXME: getReductionToWrite() returns a concat'd set of stacked scenario params, which is
+                # serialized as a giant dataframe to parquet. move this writing to be incremental, but need to
+                # verify there are no downstream consumers of the dataframe. in the meantime, limit the amount
+                # of data we'll pin in memory
+                if len(self.reduction_params) < 50:
+                    self.reduction_params.append(sub_npi.getReductionToWrite())
+                else:
+                    warnings.warn("Only storing debug information for the first 50 reduction scenarios "
+                                  "in stacked NPI as not to exhaust memory")
```

Comment:
If there are more than 50 scenarios, do the first 50 have to be appended to reduction_params?

Reply:
no, it could be the last 50 or random 50. q: can we write these as separate files (or blocks of files) vs 1 giant dataframe?

Reply:
we definitely need all of the reduction parameters to be written.  See other comment.

Reply:
There shouldn't be a large number of stacked parameters in memory at a time, even in the case they were talking about.

Reply:
normally this will never be triggered (it's only triggered for that 1 massive scenario really); it's a release valve for memory pressure. instead of concating a giant dataframe, i was thinking of having it dump periodically to disk, but wanted to save that for later as it has downstream implications.

Reply:
i think normally we have more than 50 interventions in a single stack though

Reply:
got it, we can safely up the cap, what's a reasonable max? 250? (fundamentally the approach is unscalable so the cap is temporary until it can fixed)

Reply:
250 seems ok i guess. can we change the warning to an error?

Reply:
oh, wait, that's not practical here.  can we throw an error if you try to write reductions to file and they've exceeded the cap?

Reply:
we are always writing reductions to a file regardless if it's an inference run. also, if i error out, it means we can't run a very large stacked config in inference mode.

how about we don't write reductions if the cap is exceeded? this way you don't get partial data. (bandaid until we write this data incrementally)

Reply:
i pushed an update that doesn't write reductions (it writes an error) if the cap is exceeded.

Reply:
sorry, missed these earlier.  I think either of the above is fine, as long as it doesn't write a partial config to file without error.

Reply:
got it, take a look at the latest commit, it will write an error (no partial writes)